src/zjs_fs.c

```diff
@@ -0,0 +1,1043 @@
+        mode = MODE_WX_PLUS;
+    } else if (strcmp(str, "a") == 0) {
+        mode = MODE_A;
+    } else if (strcmp(str, "ax") == 0) {
+        mode = MODE_AX;
+    } else if (strcmp(str, "a+") == 0) {
+        mode = MODE_A_PLUS;
+    } else if (strcmp(str, "ax+") == 0) {
+        mode = MODE_AX_PLUS;
+    }
+    return mode;
+}
+
+static jerry_value_t is_file(const jerry_value_t function_obj,
+                             const jerry_value_t this,
+                             const jerry_value_t argv[],
+                             const jerry_length_t argc)
+{
+    struct fs_dirent* entry;
+
+    if (!jerry_get_object_native_handle(this, (uintptr_t*)&entry)) {
+        return zjs_error("native handle not found");
+    }
+    if (entry->type == FS_DIR_ENTRY_FILE) {
+        return jerry_create_boolean(true);
+    } else {
+        return jerry_create_boolean(false);
+    }
+}
+
+static jerry_value_t is_directory(const jerry_value_t function_obj,
+                                  const jerry_value_t this,
+                                  const jerry_value_t argv[],
+                                  const jerry_length_t argc)
+{
+    struct fs_dirent* entry;
+
+    if (!jerry_get_object_native_handle(this, (uintptr_t*)&entry)) {
+        return zjs_error("native handle not found");
+    }
+    if (entry->type == FS_DIR_ENTRY_DIR) {
+        return jerry_create_boolean(true);
+    } else {
+        return jerry_create_boolean(false);
+    }
+}
+
+void free_stats(const uintptr_t native)
+{
+    struct zfs_dirent* entry = (struct zfs_dirent*)native;
+    if (entry) {
+        zjs_free(entry);
+    }
+}
+
+static jerry_value_t create_stats_obj(struct fs_dirent* entry)
+{
+
+    jerry_value_t stats_obj = jerry_create_object();
+
+    struct fs_dirent* new_entry = zjs_malloc(sizeof(struct fs_dirent));
+    if (!new_entry) {
+        return zjs_error("malloc failed");
+    }
+    memcpy(new_entry, entry, sizeof(struct fs_dirent));
+
+    jerry_set_object_native_handle(stats_obj, (uintptr_t)new_entry, NULL);
+
+    zjs_obj_add_function(stats_obj, is_file, "isFile");
+    zjs_obj_add_function(stats_obj, is_directory, "isDirectory");
+
+    return stats_obj;
+}
+
+static jerry_value_t zjs_open(const jerry_value_t function_obj,
+                              const jerry_value_t this,
+                              const jerry_value_t argv[],
+                              const jerry_length_t argc,
+                              uint8_t async)
+{
+    if (argc < 3 && async) {
+        return zjs_error("not enough arguments");
+    }
+    if (!jerry_value_is_string(argv[0])) {
+        return zjs_error("first parameter must be path string");
+    }
+    if (!jerry_value_is_string(argv[1])) {
+        return zjs_error("second parameter must be flags string");
+    }
+    if (async) {
+        if (!jerry_value_is_function(argv[argc - 1])) {
+            return zjs_error("last parameter must be callback function");
+        }
+    }
```

Comment:
These are a lot of big strings, all three could probably just be "invalid argument" and they can look up the details.